### PR TITLE
coll/xhc: Bring in latest developments

### PR DESCRIPTION
**Pull request by gkatev**
Multiple changes, new features, and enchancements

Notable new features:

- Inlining data in control path for extra small messages.
- Allreduce goes through Bcast's implementation intead of duplicating it.
- Fine grained op-wise tuning for hierarchy, chunk size, cico threshold.
- Reduce reworked & improved, with double buffering and other optimizations.
- When smsc support is not present, handle only small messages (below cico threshold), instead of not at all.

### CODE REVIEW

File: ompi/mca/coll/xhc/README.md [hunk 1]
```diff
@@ -70,213 +85,287 @@ hierarchy and its configuration.
...
+suite](https://mvapich.cse.ohio-state.edu/benchmarks/). However, when
+micro-benchmarking intra-node collectives, there are some important details
+that one needs to look out for.
 
-Computer Architecture and VLSI Systems (CARV) Laboratory, ICS Forth
+**CPU Cache** An issue with the OSU micro-benchmarks is that they use the same
+buffer for each iteration without altering it. Since modern processors
+implicitly cache data, this can lead to false/unrealistic/unrepresentative
+results, given that actual real-world applications do not (usually/optimally!)
+perform duplicate operations.
```
Comment (arun-chandran-edarath) [inline ompi/mca/coll/xhc/README.md:293]: I've also observed that using empty buffers doesn't provide accurate latency measurements while benchmarking nt-buffer-transfer (Ref: https://github.com/openucx/ucx/pull/9408) in intra-node use cases.

I appreciate you sharing the modified version of OSU-benchmarks, I will try this out.

I noticed another aspect about OSU-benchmarks, which always use page-aligned buffers for latency measurement. This may not accurately represent real-world MPI applications that might use unaligned transfers, potentially suffering from different penalties and even substantially decreased performance on some AMD CPUs (https://lunnova.dev/articles/ryzen-slow-short-rep-mov/). Have you encountered similar findings in your experience?

Comment (arun-chandran-edarath) [reply ompi/mca/coll/xhc/README.md:293]: @gkatev Do you have a table with the performance improvement numbers for any HPC apps (or any new AI apps) with xHC for intra-node cases with details like ranksXthreads, mapping, binding, machine CPU details, etc ?

Comment (gkatev) [reply ompi/mca/coll/xhc/README.md:293]: Have you had a look in our paper "A framework for hierarchical single-copy MPI collectives on multicore nodes" @ Cluster 2022? It includes a few HPC app benchmarks with XHC. But there have been many improvements to XHC between the version in that paper and the one in this PR.

Comment (arun-chandran-edarath) [reply ompi/mca/coll/xhc/README.md:293]: I went through the paper and I see you are getting good improvements with PiSvM, miniAMR and CNTK. Have you had a chance to test the performance on WRF, NWCHEM, or OpenFOAM?

Comment (gkatev) [reply ompi/mca/coll/xhc/README.md:293]: Of those I've only tried OpenFOAM at some point. If I recall correctly, it didn't make all that vast use of collectives like Allreduce, Reduce, Bcast, Barrier. Do you know how heavy use of collectives WRF and NWCHEM do?

Comment (arun-chandran-edarath) [reply ompi/mca/coll/xhc/README.md:293]: The collective percentage should depend on the type of input used right? I don't have much idea about a particular input and the MPI percentage/individual collective percentage, an application expert might help I guess :)

**Comment by gkatev** [comment]
I assume the mpi4py CI failure here is unrelated to this PR.

**Comment by hppritcha** [comment]
No this is due to a known race condition issue unrelated to this pr.  See my latest pr against main which addresses the race condition.

File: ompi/mca/coll/xhc/coll_xhc.c [hunk 2]
```diff
@@ -9,609 +9,677 @@
...
+    module->init = true;
+
+    if((mca_coll_xhc_component.print_info & XHC_PRINT_INFO_CONFIG) && rank == 0) {
+        err = xhc_print_config_info(module, comm);
 
-        void *my_cico = xhc_shmem_create(&cico_ds,
-            OMPI_XHC_CICO_MAX, comm, "cico", 0);
-        if(!my_cico) {
-            RETURN_WITH_ERROR(return_code, OMPI_ERR_OUT_OF_RESOURCE, end);
+        if(err != OMPI_SUCCESS) {
```
Comment (edgargabriel) [inline ompi/mca/coll/xhc/coll_xhc.c:88]: Minor: the Open MPI coding guidelines suggest that you use the constant in a comparison as a first argument, i.e. this would be

```
if (OMPI_SUCCESS != err) {
```

(this occurs at multiple locations in the code, will not point it out beyond this one)

File: ompi/mca/coll/xhc/coll_xhc.c [hunk 3]
```diff
@@ -9,609 +9,677 @@
...
+        return OMPI_SUCCESS;
+
+    xhc_module_set_coll_fns(comm, &module->prev_colls, &xhc_fns);
+
+    // --
 
-        free(data->comms);
-        free(data);
+    ds_list = malloc(comm_size * sizeof(opal_shmem_ds_t));
+    if(!ds_list) return OMPI_ERR_OUT_OF_RESOURCE;
```
Comment (edgargabriel) [inline ompi/mca/coll/xhc/coll_xhc.c:238]: Minor: (sorry to bring up again an Open MPI coding guideline). Even single line blocks should be encapsulated in a pair of {}

**Review by edgargabriel — CHANGES_REQUESTED** [review]

File: ompi/mca/coll/xhc/coll_xhc.c (hunk 2 above)
Comment (gkatev) [reply ompi/mca/coll/xhc/coll_xhc.c:88]: Hmm yes, I'm not too proud of my non-compliance here. But this rule is really hard to follow. The equality check is not just that, the code tells a story, and it tells it left-to-right (at least to me). So when I see `x == 0` I read it as if `x` has the value of 0, whereas in `0 == x`, `0` has the value of `x`?? And this is kind of a simple case, it can get more complex. Furthermore, this rule often leads to reversing the operands in relational operations, e.g. things like `0 < count`, where all hell breaks loose trying to read it. And as for the safety aspect, GCC has `-Wparentheses` which warns about this problem; we could even make this an error (in XHC? in all of OMPI?).

Of course this is not the platform to argue about the coding style, sorry for the manifesto :-). I can change all of the code to reverse the operands, but honestly I'm more worried about bugs creeping in as a result of this conversion, rather than an accidental omission of an `=` or `!`. But if it must be...

File: ompi/mca/coll/xhc/coll_xhc.c (hunk 3 above)
Comment (gkatev) [reply ompi/mca/coll/xhc/coll_xhc.c:238]: Thanks for taking the time to look into the PR. I'm less emotional about this one :-). Once again, vis-a-vis the safety aspect, I'd argue that GCC has `-Wmisleading-indentation` in `-Wall` for this purpose. I find it useful to not always encapsulate in `{}` especially for single-line ifs, because otherwise one can end up with a lot of extra lines, making functions longer and harder to follow. Of course now that I think of it, one could still have the braces on a single line (a bit unusual, but certainly doable, and perhaps better than all the extra lines).

I avoided converting these occurrences to make it easier to compare and keep track of changes between my development repo and the upstream version, to make sure I don't miss anything. But once again, if it must be, I can put in the work to do these conversions.

File: ompi/mca/coll/xhc/coll_xhc.c (hunk 2 above)
Comment (rhc54) [reply ompi/mca/coll/xhc/coll_xhc.c:88]: Reason for this rule is very simple and compelling: if you make a typo and have `x = 0`, then that is a valid assignment and will always equate to true. However, if you have `0 = x`, that is an invalid assignment and will generate a compile error. So it is a simple and effective way to protect against mistakes.

Comment (gkatev) [reply ompi/mca/coll/xhc/coll_xhc.c:88]: Sure I understand that part. IMHO it probably still doesn't outweigh the readability problem. But either way, isn't this typo nowadays easily caught with `-Wparentheses` (or even `-Werror=parentheses`)? Is there still a reason to do it manually, given the readability disadvantage of `0 = x`?

Comment (rhc54) [reply ompi/mca/coll/xhc/coll_xhc.c:88]: Well, readability is a pretty subjective thing and we may not all agree with your statement. I don't know when that warning showed up, but definitely not in older compilers and perhaps in some we still support (we just finally turned off gcc6!), and I don't know if other compiler families support it. Besides, there is a tendency over here for people to ignore warnings and to strongly resist treating warnings as errors, so relying on a compiler warning to protect us is probably a losing proposition.

Given how many times this simple rule has saved our tails, it is rather difficult to justify removing it. Still, you are welcome to try and convince the community to do so!

**Comment by edgargabriel** [comment]
I would honestly look at it slightly differently, we can discuss these issues ad absurdum, in my opinion they are not earth shaking either way (that's why I marked them as minor). The same rules are not being used in a number of other projects that I am involved, so I truly do not hold any religious beliefs in them.

However, I do see a benefit for a multi-institutional project such as Open MPI to follow some coding guidelines, to avoid that different parts of the code look vastly different.

File: ompi/mca/coll/xhc/coll_xhc.c (hunk 2 above)
Comment (gkatev) [reply ompi/mca/coll/xhc/coll_xhc.c:88]: That's fair enough. Personally, I'd rather rely on the compiler for this, especially given the alternative, and given that in the XHC codebase I don't leave warnings around.

For some completeness, since we discussed it, a quick test reveals that the warning is present at least as far back as GCC 4.8 with `-Wall`, and it's also on the latest Clang (by default). It's also present on some ICC I tested, but only with `-Wparentheses`, not with just `-Wall`. Of course this is not exhaustive, and indeed this is not the appropriate place to figure it out either...

**Comment by gkatev** [comment]
I agree it's not the end of the world; I did kind of go and open a can of worms here, and I can't reasonably argue against adhering to an agreed upon standard. There are some issues associated with doing these conversions, so I *would* avoid them if others agreed that the specific guideline(s) are not as heavily enforced/required/necessary (??), but it certainly can be done

**Comment by rhc54** [comment]
Guess I'm more aligned with Edgar on this one - not religious about it, but there is a benefit to having coding standards, and it makes little sense to have them if we start saying "well not for this PR". 🤷‍♂️  Ultimately up to the community.

**Comment by gkatev** [comment]
That's understandable -- I'll do the conversion, thanks for the input

**Comment by gkatev** [comment]
I've rectified the coding style issues, this is again ready for review.

Did you maybe have any other comments for this PR? @bosilca did you maybe have a chance to look this over? (sorry for the 'explicit ping', it appears I can't 'request a review' here).

**Review by edgargabriel — APPROVED** [review]
LGTM

**Review by arun-chandran-edarath — APPROVED** [review]
Looks Good.
